### ISSUE TITLE
Periodically strange, every time new, results. Each new training network process.

### ISSUE THREAD

**Issue by lubagov** [BVLC/caffe#6376]
### Issue summary
Hello! I try to learn Caffe, it is my first steps, to use this framework.

I build this example from here, https://github.com/shi-yan/caffe/tree/master/examples/cpp_xor , together with last commit 864520713a4c5ffae7382ced5d34e4cadc608473 Caffe  and try to test it. But result so strange. Periodically network fine trained, and i get good results, periodically not.
I use CPU for train this network, but together with CuBLAS or MKL, result same.
Also i try to use Caffe 1.0 release and get same result.

In 30% cases i get this result, network training, in 70% all is ok:
```
input: 0 xor 0,  truth: 0.000000 result by nn: 0.000287
input: 0 xor 1,  truth: 1.000000 result by nn: 0.999608
input: 1 xor 0,  truth: 1.000000 result by nn: 0.510007
input: 1 xor 1,  truth: 0.000000 result by nn: 0.510325
```
I am also try to modify this sample, and try train ResNet34 and I get a similar error.
Sometimes it works well, sometimes it does not.....

Also i have some problem, when i train network in GPU, and try use it in CPU mode. I get nonsense. Meanwhile, if testing and use is performed on the same device, then the result is good.
Probably I just do something wrong, but it does not look clear.
I run unit tests, but some get falue:
```
[----------] Global test environment tear-down
[==========] 2199 tests from 285 test cases ran. (363610 ms total)
[  PASSED  ] 2195 tests.
[  FAILED  ] 4 tests, listed below:
[  FAILED  ] NetTest/0.TestFromTo, where TypeParam = caffe::CPUDevice<float>
[  FAILED  ] NetTest/1.TestFromTo, where TypeParam = caffe::CPUDevice<double>
[  FAILED  ] BatchReindexLayerTest/2.TestGradient, where TypeParam = caffe::GPUDevice<float>
[  FAILED  ] BatchReindexLayerTest/3.TestGradient, where TypeParam = caffe::GPUDevice<double>
```

### System configuration

* Operating system: Ubuntu 16.04
* Compiler:  g++ (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
* CUDA version (if applicable):  9.1.85-1
* CUDNN version (if applicable):  7.0.5.15-1+cuda9.1
* BLAS: MKL 2018.2.199, CuBLAS 9.1.85-1
* GPU: Nvidia GT1030

I build second configuration, with blas, ATLAS 3.10.2 and without CUDA. Now all test passed, but this sample same working, periodically good, periodically not good...

**Comment by lubagov** [comment]
I try to create LMDB database, and next use it for train in caffe:
```../build/tools/caffe train --solver=./solver.prototxt```
I sqme get starnge result:
test set is:
```
0 xor 0 = 0
0 xor 1 = 1
1 xor 0 = 1
1 xor 1 = 0
```
```
I0430 14:32:40.691485 12390 solver.cpp:418]     Test net output #0: loss = 0.347155 (* 1 = 0.347155 loss)
I0430 14:32:40.691493 12390 solver.cpp:418]     Test net output #1: output = 0.000295797
I0430 14:32:40.691530 12390 solver.cpp:418]     Test net output #2: output = 0.999695
I0430 14:32:40.691540 12390 solver.cpp:418]     Test net output #3: output = 0.483335
I0430 14:32:40.691556 12390 solver.cpp:418]     Test net output #4: output = 0.483652
```
but sometimes, it good working:
```
I0430 14:42:32.581300 12492 solver.cpp:351] Iteration 5000000, Testing net (#0)
I0430 14:42:32.581306 12492 net.cpp:676] Ignoring source layer inputdata
I0430 14:42:32.581331 12492 solver.cpp:418]     Test net output #0: loss = 0.00057354 (* 1 = 0.00057354 loss)
I0430 14:42:32.581357 12495 data_layer.cpp:73] Restarting data prefetching from start.
I0430 14:42:32.581357 12492 solver.cpp:418]     Test net output #1: output = 0.000560813
I0430 14:42:32.581379 12492 solver.cpp:418]     Test net output #2: output = 0.999339
I0430 14:42:32.581388 12492 solver.cpp:418]     Test net output #3: output = 0.999419
I0430 14:42:32.581395 12492 solver.cpp:418]     Test net output #4: output = 0.000490535
```
although, everything should be the same... every training, according to the logic of things...

i attach my net and dataset:
[testlmdb.tar.gz](https://github.com/BVLC/caffe/files/1960406/testlmdb.tar.gz)

**Comment by Noiredd** [comment]
It's normal that the network will learn slightly differently every time, as the layers are initialized to different random values each time. We can't help you with detailed troubleshooting, as the code example you are asking about is not a part of Caffe but a fork by someone else.  
You can consider asking this question again on the [caffe-users](https://groups.google.com/forum/#!forum/caffe-users) group.
